File: apps/cli/src/Haas.Hosty.Cli/Mcp/ToolKey.cs
namespace Haas.Hosty.Cli.Mcp;

using System.Security.Cryptography;
using System.Text;

/// <summary>
/// Maps an app's tool onto the name an MCP client will show the model, per the mapping settled in
/// docs/features/hosty-mcp-connector/plan.md.
/// </summary>
/// <remarks>
/// The exported name has to be <b>unique and stable</b> — not decodable. The connector keeps its own
/// table from exported name back to (app, interface, tool), so nothing ever parses these strings; what
/// matters is that two distinct tools never collide, and that a given tool's name does not change when
/// some <i>other</i> app is installed. The second property is why collisions are avoided by
/// construction rather than by disambiguating after the fact: a client's permission rules are written
/// against these names, and a name that shifts under them fails silently.
/// </remarks>
internal static class ToolKey
{
    /// <summary>The interface key that is left out of the name, since nearly every app uses it.</summary>
    public const string DefaultInterfaceKey = "default";

    /// <summary>
    /// Longest key (app + interface) before it is replaced by a truncated form. Core accepts an app id
    /// of up to 63 characters and the escape below can nearly double one, so an unbounded key would
    /// push whole apps past what clients accept — see rule 3 in the plan.
    /// </summary>
    private const int MaxKeyChars = 32;

    /// <summary>Characters kept from the digest when a key is truncated: 32 bits of app-specific hash.</summary>
    private const int DigestChars = 8;

    /// <summary>
    /// Default ceiling on the name this connector exports. A client prepends its own
    /// <c>mcp__&lt;server&gt;__</c>, which is 12 characters for a server entry named <c>hosty</c>, so
    /// this default keeps the string the model finally sees within 64. Packaging that uses longer
    /// server names should lower it.
    /// </summary>
    public const int DefaultMaxToolNameChars = 52;

    /// <summary>
    /// The stable key for one app interface. Pure in its two arguments — nothing about the rest of the
    /// fleet reaches it.
    /// </summary>
    public static string ForInterface(string appId, string? interfaceKey)
    {
        var key = string.IsNullOrWhiteSpace(interfaceKey) ? DefaultInterfaceKey : interfaceKey.Trim();
        var escaped = Escape(appId);
        var composed = string.Equals(key, DefaultInterfaceKey, StringComparison.Ordinal)
            ? escaped
            : $"{escaped}__{key}";

        return composed.Length <= MaxKeyChars ? composed : Truncate(composed, appId, key);
    }

    /// <summary>
    /// The exported tool name, or null when the tool cannot be offered under this mapping.
    /// </summary>
    /// <remarks>
    /// Two refusals, both of which the caller logs rather than swallowing:
    /// <list type="bullet">
    /// <item>a tool whose own name contains <c>__</c>, which would make the boundary between key and
    /// tool ambiguous — this is what stops app <c>X</c>'s tool <c>admin__foo</c> colliding with app
    /// <c>X</c>'s <c>admin</c> interface offering <c>foo</c>;</item>
    /// <item>a name over the ceiling. Refused, never truncated: truncating collides tool names with
    /// each other, which is strictly worse than not offering one.</item>
    /// </list>
    /// </remarks>
    public static string? ForTool(string key, string toolName, int maxToolNameChars = DefaultMaxToolNameChars)
    {
        if (string.IsNullOrWhiteSpace(toolName) || toolName.Contains("__", StringComparison.Ordinal))
        {
            return null;
        }

        var name = $"{key}__{toolName}";
        return name.Length <= maxToolNameChars ? name : null;
    }

    /// <summary>
    /// Escapes an app id into <c>[a-z0-9_-]</c> such that the result can never contain <c>__</c>:
    /// every <c>_</c> it emits is immediately followed by <c>d</c>, <c>u</c>, or <c>x</c>. That is the
    /// property the whole mapping rests on, because it makes the first <c>__</c> in a composed name an
    /// unambiguous boundary.
    /// </summary>
    /// <remarks>
    /// Reversible, which is how it stays injective — the naive alternative of replacing <c>.</c> with
    /// <c>-</c> maps <c>com.example.notes</c> and <c>com-example-notes</c> onto the same string, a bug
    /// already found and fixed once in the AI gateway.
    /// Core's id pattern admits only <c>[a-z0-9._-]</c>, so the <c>_x</c> escape is unreachable through
    /// a validated manifest; it exists so a hand-edited registry cannot produce a colliding key.
    /// </remarks>
    internal static string Escape(string appId)
    {
        var builder = new StringBuilder(appId.Length + 8);
        foreach (var character in appId)
        {
            switch (character)
            {
                case '.':
                    builder.Append("_d");
                    break;
                case '_':
                    builder.Append("_u");
                    break;
                case >= 'a' and <= 'z':
                case >= '0' and <= '9':
                case '-':
                    builder.Append(character);
                    break;
                default:
                    builder.Append("_x").Append(((int)character).ToString("x2"));
                    break;
            }
        }

        return builder.ToString();
    }

    /// <summary>
    /// Replaces an over-long key with a fixed-width form: a readable prefix, then a digest of the
    /// identity it came from. The digest is taken over the <i>unescaped</i> app id and interface key
    /// joined by a space — a character neither may contain — so two different identities cannot feed
    /// the hash the same bytes.
    /// </summary>
    private static string Truncate(string composed, string appId, string interfaceKey)
    {
        var digest = Convert.ToHexStringLower(
            SHA256.HashData(Encoding.UTF8.GetBytes($"{appId} {interfaceKey}")))[..DigestChars];
        return $"{composed[..(MaxKeyChars - DigestChars - 1)]}-{digest}";
    }
}

File: apps/ai-gateway/src/facade/tool-key.ts
// The name an aggregated tool is offered under.
//
// A deliberate port of the CLI connector's mapping (apps/cli/…/Mcp/ToolKey.cs, settled in
// docs/features/hosty-mcp-connector/feature.md) rather than a second scheme. Both surfaces present
// the same fleet to the same clients, and a client's permission rules are written against these
// strings — two dialects would mean a rule that works through `hosty mcp` and silently does not
// through the facade. The port is asserted against the connector's own worked examples in the tests.
//
// The exported name has to be **unique and stable**, not decodable: nothing ever parses these back,
// because the facade keeps its own table from name to (app, tool). Stability is why collisions are
// avoided by construction rather than by disambiguating after the fact — a name that shifts when
// some *other* app is installed breaks a client rule without any visible failure.

import { createHash } from "node:crypto";

/** The interface key left out of a name, since nearly every app uses it. */
export const DEFAULT_INTERFACE_KEY = "default";

/**
 * Longest key (app + interface) before it is replaced by a truncated form. Core accepts an app id of
 * up to 63 characters and the escape below can nearly double one, so an unbounded key would push
 * whole apps past what clients accept.
 */
const MAX_KEY_CHARS = 32;

/** Characters kept from the digest when a key is truncated: 32 bits of app-specific hash. */
const DIGEST_CHARS = 8;

/**
 * Ceiling on the exported name. A client prepends its own `mcp__<server>__`, so this keeps what the
 * model finally sees within 64 for a server entry named `hosty`.
 */
export const DEFAULT_MAX_TOOL_NAME_CHARS = 52;

/** The stable key for one app interface. Pure in its arguments — nothing about the rest of the
 * fleet reaches it, which is the property that keeps names from shifting. */
export function interfaceKey(appId: string, key: string = DEFAULT_INTERFACE_KEY): string {
  const resolved = key.trim() || DEFAULT_INTERFACE_KEY;
  const escaped = escapeAppId(appId);
  const composed = resolved === DEFAULT_INTERFACE_KEY ? escaped : `${escaped}__${resolved}`;
  if (composed.length <= MAX_KEY_CHARS) {
    return composed;
  }

  // The digest is taken over the *unescaped* identity joined by a space — a character neither part
  // may contain — so two different identities cannot feed the hash the same bytes.
  const digest = createHash("sha256").update(`${appId} ${resolved}`).digest("hex").slice(0, DIGEST_CHARS);
  return `${composed.slice(0, MAX_KEY_CHARS - DIGEST_CHARS - 1)}-${digest}`;
}

/**
 * The exported tool name, or null when the tool cannot be offered under this mapping.
 *
 * Two refusals, both of which the caller logs rather than swallowing: a tool whose own name contains
 * `__`, which would make the boundary between key and tool ambiguous; and a name over the ceiling,
 * refused rather than truncated, because truncating collides tool names with each other — strictly
 * worse than not offering one.
 */
export function toolName(
  key: string,
  tool: string,
  maxToolNameChars: number = DEFAULT_MAX_TOOL_NAME_CHARS,
): string | null {
  if (!tool.trim() || tool.includes("__")) {
    return null;
  }

  const name = `${key}__${tool}`;
  return name.length <= maxToolNameChars ? name : null;
}

/**
 * Escapes an app id into `[a-z0-9_-]` such that the result can never contain `__`: every `_` it
 * emits is immediately followed by `d`, `u`, or `x`. That is the property the whole mapping rests
 * on, because it makes the first `__` in a composed name an unambiguous boundary.
 *
 * Reversible, which is how it stays injective — the naive alternative of replacing `.` with `-` maps
 * `com.example.notes` and `com-example-notes` onto the same string, a bug this repository has
 * already found and fixed once.
 */
export function escapeAppId(appId: string): string {
  let out = "";
  for (const character of appId) {
    if (character === ".") {
      out += "_d";
    } else if (character === "_") {
      out += "_u";
    } else if (/[a-z0-9-]/.test(character)) {
      out += character;
    } else {
      // Unreachable through a validated manifest (Core's id pattern admits only [a-z0-9._-]); it
      // exists so a hand-edited registry cannot produce a colliding key.
      out += `_x${character.codePointAt(0)!.toString(16).padStart(2, "0")}`;
    }
  }
  return out;
}
